README.md
[![Build Status](https://travis-ci.org/kfly8/p5-Function-Interface.svg?branch=master)](https://travis-ci.org/kfly8/p5-Function-Interface) [![Coverage Status](https://img.shields.io/coveralls/kfly8/p5-Function-Interface/master.svg?style=flat)](https://coveralls.io/r/kfly8/p5-Function-Interface?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/Function-Interface.svg)](https://metacpan.org/release/Function-Interface)
# NAME

Function::Interface - declare typed interface package

# SYNOPSIS

Declare typed interface package `IFoo`:

```perl
package IFoo {
    use Function::Interface;
    use Types::Standard -types;

    fun hello(Str $msg) :Return(Str);

    fun add(Int $a, Int $b) :Return(Int);
}
```

Implements the interface package `IFoo`:

```perl
package Foo {
    use Function::Interface::Impl qw(IFoo);
    use Types::Standard -types;

    fun hello(Str $msg) :Return(Str) {
        return "HELLO $msg";
    }

    fun add(Int $a, Int $b) :Return(Int) {
        return $a + $b;
    }
}
```

# DESCRIPTION

This module provides a typed interface.
`Function::Interface` declares abstract functions without implementation and defines an interface package.
`Function::Interface::Impl` checks if the abstract functions are implemented at **compile time**.

## SUPPORT

This module supports all perl versions starting from v5.14.

## Declare function

`Function::Interface` provides two new keywords, `fun` and `method`, for declaring abstract functions and methods with types:

```
fun hello(Str $msg) :Return(Str);

method new(Num :$x, Num :$y) :Return(Point);
```

The method of declaring abstract functions is the same as [Function::Parameters](https://metacpan.org/pod/Function::Parameters) and [Function::Return](https://metacpan.org/pod/Function::Return).

### declare parameters

Function arguments must always specify a variable name and type constraint, and named arguments and optional arguments can optionally be specified:

```perl
# positional parameters
# e.g. called `foo(1,2,3)`
fun foo1(Int $a, Int $b, Int $c) :Return();

# named parameters
# e.g. called `bar(x => 123, y => 456)`
fun foo2(Num :$x, Num :$y) :Return();

# optional
# e.g. called `baz()` or `baz('some')`
fun foo3(Str $msg=) :Return();
```

### declare return types

Specify zero or more type constraints for the function's return value:

```
# zero(empty)
fun bar1() :Return();

# one
fun bar2() :Return(Str);

# two
fun bar3() :Return(Str, Num);
```

## requirements of type constraint

The requirements of type constraint of `Function::Interface` is the same as for [Function::Parameters](https://metacpan.org/pod/Function::Parameters) and [Function::Return](https://metacpan.org/pod/Function::Return).

# METHODS

## Function::Interface::info($interface\_package)

The function `Function::Interface::info` lets you introspect interface functions:

```perl
# declare interface package
package IBar {
    use Function::Interface;
    fun hello() :Return();
    fun world() :Return();
}

# introspect
my $info = Function::Interface::info 'IBar';
$info->package; # => IBar
$info->functions; # => list of Function::Interface::Info::Function
```

It returns either `undef` if it knows nothing about the interface or an object of [Function::Interface::Info](https://metacpan.org/pod/Function::Interface::Info).

# SEE ALSO

[Function::Interface::Impl](https://metacpan.org/pod/Function::Interface::Impl)

# LICENSE

Copyright (C) kfly8.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

kfly8 <kfly@cpan.org>


Changes
Revision history for Perl extension Function-Interface

{{$NEXT}}

0.06 2019-06-19T15:28:30Z
    - use B::Hooks::EndOfScope instead of CHECK phase #11

0.05 2019-03-20T23:29:16Z
    - fixed pod. thanks @manwar! #10

0.04 2019-03-20T14:49:19Z
    - [feature] automatically import Function::Parameters and Function::Return
    - add docs and tests 
    - refactor

0.03 2019-03-15T15:08:52Z
    - add tests

0.02 2019-03-12T16:33:06Z
    - [BUGFIX] rename Types::Interface to Function::Interface::Types

0.01 2019-03-01T03:53:13Z

    - original version



t/lib/FooNoParamsInfo.pm
package FooNoParamsInfo;
use Function::Interface::Impl qw(IFoo);
use Function::Return;

sub foo :Return() {}

1;


t/lib/FooNoReturnInfo.pm
package FooNoReturnInfo;
use Function::Interface::Impl qw(IFoo);
use Function::Parameters;

fun foo() {}

1;


t/lib/FooInvalidParams.pm
package FooInvalidParams;
use Function::Interface::Impl qw(IFoo);
use Function::Return;
use Function::Parameters;

use Types::Standard -types;

fun foo(Str $a) :Return() {}

1;


lib/Function/Interface/Impl.pm
package Function::Interface::Impl;

use v5.14.0;
use warnings;

our $VERSION = "0.05";

use Class::Load qw(try_load_class is_class_loaded);
use Scalar::Util qw(blessed);
use Import::Into;
use B::Hooks::EndOfScope;

use Function::Interface;
use Function::Parameters;
use Function::Return;

sub import {
    my $class = shift;
    my @interface_packages = @_;
    my ($pkg, $filename, $line) = caller;

    for (@interface_packages) {
        _register_check_list($pkg, $_, $filename, $line);
    }

    Function::Parameters->import::into($pkg);
    Function::Return->import::into($pkg);

    on_scope_end {
        _check_impl();
    }
}

our @CHECK_LIST;
my %IMPL_CHECKED;
sub _check_impl {
    while (my $data = shift @CHECK_LIST) {
        my ($package, $interface_package, @fl) = @$data{qw/package interface_package filename line/};
        assert_valid($package, $interface_package, @fl);

        # for Function::Interface::Types#ImplOf
        $IMPL_CHECKED{$package}{$interface_package} = !!1;
    }
}

sub _register_check_list {
    my ($package, $interface_package, $filename, $line) = @_;

    push @CHECK_LIST => +{
        package           => $package,
        interface_package => $interface_package,
        filename          => $filename,
        line              => $line,
    }
}

sub assert_valid {
    my ($package, $interface_package, $filename, $line) = @_;
    my @fl = ($filename, $line);

    {
        my $ok = is_class_loaded($package);
        return _error("implements package is not loaded yet. required to use $package", @fl) if !$ok;
    }

    {
        my ($ok, $e) = try_load_class($interface_package);
        return _error("cannot load interface package: $e", @fl) if !$ok;
    }

    my $iinfo = info_interface($interface_package)
            or return _error("cannot get interface info", @fl);

    for my $ifunction_info (@{$iinfo->functions}) {
        my $fname = $ifunction_info->subname;
        my $def   = $ifunction_info->definition;

        my $code = $package->can($fname)
            or return _error("function `$fname` is required. Interface: $def", @fl);

        my $pinfo = info_params($code)
            or return _error("cannot get function `$fname` parameters info. Interface: $def", @fl);
        my $rinfo = info_return($code)
            or return _error("cannot get function `$fname` return info. Interface: $def", @fl);

        check_params($pinfo, $ifunction_info)
            or return _error("function `$fname` is invalid parameters. Interface: $def", @fl);
        check_return($rinfo, $ifunction_info)
            or return _error("function `$fname` is invalid return. Interface: $def", @fl);
    }
}

sub _error {
    my ($msg, $filename, $line) = @_;
    die sprintf "implements error: %s at %s line %s\n\tdied", $msg, $filename, $line;
}

sub info_interface {
    my $interface_package = shift;
    Function::Interface::info($interface_package)
}

sub info_params {
    my $code = shift;
    Function::Parameters::info($code)
}

sub info_return {
    my $code = shift;
    Function::Return::info($code)
}

sub check_params {
    my ($pinfo, $ifunction_info) = @_;

    return unless $ifunction_info->keyword eq $pinfo->keyword;

    my $params_count = 0;
    for my $key (qw/positional_required positional_optional named_required named_optional/) {
        my @params = $pinfo->$key;
        $params_count += @params;

        for my $i (0 .. $#{$ifunction_info->$key}) {
            my $ifp = $ifunction_info->$key->[$i];
            my $p = $params[$i];
            return unless check_param($p, $ifp);
        }
    }

    return unless $params_count == @{$ifunction_info->params};
    return !!1
}

sub check_param {
    my ($param, $iparam) = @_;
    return unless $param;
    return $iparam->type eq $param->type
        && $iparam->name eq $param->name
}

sub check_return {
    my ($rinfo, $ifunction_info) = @_;

    return unless @{$rinfo->types} == @{$ifunction_info->return};

    for my $i (0 .. $#{$ifunction_info->return}) {
        my $ifr = $ifunction_info->return->[$i];
        my $type = $rinfo->types->[$i];
        return unless $ifr->type eq $type;
    }
    return !!1;
}

sub impl_of {
    my ($package, $interface_package) = @_;
    $package = ref $package ? blessed($package) : $package;
    $IMPL_CHECKED{$package}{$interface_package}
}

1;
__END__

=encoding utf-8

=head1 NAME

Function::Interface::Impl - implements interface package

=head1 SYNOPSIS

Implements the interface package C<IFoo>:

    package Foo {
        use Function::Interface::Impl qw(IFoo);
        use Types::Standard -types;

        fun hello(Str $msg) :Return(Str) {
            return "HELLO $msg";
        }

        fun add(Int $a, Int $b) :Return(Int) {
            return $a + $b;
        }
    }

=head1 DESCRIPTION

Function::Interface::Impl is for implementing interface package.
This module checks if the abstract functions are implemented at B<compile time> and imports Function::Parameters and Function::Return into the implementing package.

=head1 NOTES

Function::Interface must be loaded B<before> Function::Return.

You need to call C<CHECK> code blocks in the following order:
1. Function::Return#CHECK (to get return info)
2. Function::Interface::Impl#CHECK (to check implements)

C<CHECK> code blocks are LIFO order.
So, it is necessary to load in the following order:
1. Function::Interface::Impl
2. Function::Return

=head1 METHODS

=head2 assert_valid

check if the interface package is implemented, otherwise die.

=head2 info_interface($interface_package)

get the object of Function::Interface::Info.

=head2 info_params($code)

get the object of Function::Parameters.

=head2 info_return($code)

get the object of Function::Return.

=head2 check_params($params_info, $interface_function_info)

check if the arguments are implemented according to the interface info.

=head2 check_param($param, $interface_param)

check if the argument are implemented according to the interface info.

=head2 check_return($return_info, $interface_function_info)

check if the return types are implemented according to the interface info.

=head2 impl_of($package, $interface_package)

check if specified package is an implementation of specified interface package.

=head1 SEE ALSO

L<Function::Parameters>, L<Function::Return>

=head1 LICENSE

Copyright (C) kfly8.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

kfly8 E<lt>kfly@cpan.orgE<gt>

=cut



lib/Function/Interface/Info.pm
package Function::Interface::Info;

use v5.14.0;
use warnings;

our $VERSION = "0.03";

sub new {
    my ($class, %args) = @_;
    bless {
        package => $args{package},
        functions => $args{functions},
    } => $class;
}

sub package() { $_[0]->{package} }
sub functions() { $_[0]->{functions} }

1;
__END__

=encoding utf-8

=head1 NAME

Function::Interface::Info - information about interface package

=head1 SYNOPSIS

    package IFoo {
        use Function::Interface;

        fun hello(Str $msg) :Return(Str);
    }

    my $info = Function::Interface::info 'IFoo';
    $info->package;   # IFoo
    $info->functions; # [ Function::Interface::Info::Function ]

    for my $finfo (@{$info->functions}) {
        $finfo->subname; # hello
        $finfo->keyword; # fun
        $finfo->params;  # [ Function::Interface::Info::Function::Param ]
        $finfo->return;  # [ Function::Interface::Info::Function::ReturnParam ]

        for my $pinfo (@{$finfo->params}) {
            $pinfo->type;     # Str
            $pinfo->name;     # $msg
            $pinfo->named;    # false
            $pinfo->optional; # false
        }

        for my $rinfo (@{$rinfo->return}) {
            $rinfo->type; # Str
        }
    }

=head1 DESCRIPTION

Function::Interface::info returns objects of this class to describe interface package.

=head1 METHODS

=head2 new

Constructor of Function::Interface::Info. This is usually called at Function::Interface::info.

=head2 $info->package

Returns interface package name

=head2 $info->functions

Returns a list of L<Function::Interface::Info::Function>

=head1 SEE ALSO

L<Function::Interface>



lib/Function/Interface/Types.pm
package Function::Interface::Types;

use v5.14.0;
use warnings;

our $VERSION = "0.03";

use Type::Library -base,
    -declare => qw( ImplOf );

use Types::Standard -types;
use Class::Load qw(is_class_loaded);
use Scalar::Util qw(blessed);

__PACKAGE__->add_type({
    name   => 'ImplOf',
    parent => ClassName | Object,
    constraint_generator => sub {
        my $self = $Type::Tiny::parameterize_type;
        my @interfaces = @_;

        my $display_name = $self->name_generator->($self, @interfaces);

        my %options = (
            constraint   => sub {
                my ($package) = @_;
                for (@interfaces) {
                    return unless Function::Interface::Impl::impl_of($package, $_);
                }
                return !!1;
            },
            display_name => $display_name,
            parameters   => [@interfaces],
            message => sub {
                my $package = ref $_ ? blessed($_) : $_;
                return "$package is not loaded" if !is_class_loaded($package); 
                return sprintf '%s did not pass type constraint "%s"', Type::Tiny::_dd($_), $display_name;
            },
        );

        return $self->create_child_type(%options);
    },
});

__PACKAGE__->meta->make_immutable;

__END__

=encoding utf-8

=head1 NAME

Function::Interface::Types - interface types for Type::Tiny

=head1 SYNOPSIS

    use Function::Interface::Types -types;
    my $type = ImplOf['IFoo'];

    my $foo = Foo->new; # implements of 'IFoo'
    $type->check($foo); # ok

=head1 DESCRIPTION

Function::Interface::Types provides type constraints of interface package.
The following types are available.

=head2 ImplOf[`a]

C<ImplOf['IFoo', 'IBar']> allows objects or packages where implements of interface package 'IFoo' and 'IBar' both return true.

Given no parameters, just equivalent to Object or ClassName.

=head1 SEE ALSO

L<Function::Interface>


lib/Function/Interface/Info/Function.pm
package Function::Interface::Info::Function;

use v5.14.0;
use warnings;

our $VERSION = "0.04";

sub new {
    my ($class, %args) = @_;
    bless \%args => $class;
}

sub subname() { $_[0]->{subname} }
sub keyword() { $_[0]->{keyword} }
sub params() { $_[0]->{params} }
sub return() { $_[0]->{return} }

sub definition() {
    my $self = shift;

    sprintf('%s %s(%s) :Return(%s)',
        $self->keyword,
        $self->subname,
        (join ', ', map {
            sprintf('%s %s%s%s',
                $_->type_display_name,
                $_->named ? ':' : '',
                $_->name,
                $_->optional ? '=' : ''
            )
        } @{$self->params}),
        (join ', ', map {
            $_->type_display_name,
        } @{$self->return}),
    );
}

sub positional_required() {
    my $self = shift;
    $self->{positional_required} //= [ grep { !$_->named && !$_->optional } @{$self->params} ]
}

sub positional_optional() {
    my $self = shift;
    $self->{positional_optional} //= [ grep { !$_->named && $_->optional } @{$self->params} ]
}

sub named_required() {
    my $self = shift;
    $self->{named_required} //= [ grep { $_->named && !$_->optional } @{$self->params} ]
}

sub named_optional() {
    my $self = shift;
    $self->{named_optional} //= [ grep { $_->named && $_->optional } @{$self->params} ]
}

1;
__END__

=encoding utf-8

=head1 NAME

Function::Interface::Info::Function - information about abstract function of interface package

=head1 METHODS

=head2 new

Constructor of Function::Interface::Info::Function. This is usually called at Function::Interface::info.

=head2 subname

Returns an abstract function name

=head2 keyword

Returns the keyword used to define the abstract function, i.e. C<fun> or C<method>

=head2 params

Returns a list of L<Function::Interface::Info::Function::Param>

=head3 positional_required

Returns positional required params

=head3 positional_optional

Returns positional optional params

=head3 named_required 

Returns named required params

=head3 named_optional

Returns named optional params

=head2 return

Returns a list of L<Function::Interface::Info::Function::ReturnParam>

=head2 definition

Returns the abstract function declaration string. For example, "Str $msg"

=head1 SEE ALSO

L<Function::Interface::Info>



lib/Function/Interface/Info/Function/Param.pm
package Function::Interface::Info::Function::Param;

use v5.14.0;
use warnings;

our $VERSION = "0.03";

sub new {
    my ($class, %args) = @_;
    bless \%args => $class;
}

sub type() { $_[0]->{type} }
sub name() { $_[0]->{name} }
sub optional() { !!$_[0]->{optional} }
sub named() { !!$_[0]->{named} }

sub type_display_name() { $_[0]->type->display_name }

1;
__END__

=encoding utf-8

=head1 NAME

Function::Interface::Info::Function::Param - information about parameters of abstract function

=head1 METHODS

=head2 new

Constructor of Function::Interface::Info::Function::Param. This is usually called at Function::Interface::info.

=head2 name -> Str

Returns parameter name of the abstract function, e.g. C<$msg>.

=head2 named -> Bool

Returns whether it is a named parameter.
For example, C<Str $a> is false, C<Str :$b> is true.

=head2 optional -> Bool

Returns whether it is a optional parameter.
For example, C<Str $a> is false, C<Str $b=> is true.

=head2 type -> Object

Returns type object of the parameter, e.g. C<Str>.

=head3 type_display_name -> Str

Returns type display name of the parameter 

=head1 SEE ALSO

L<Function::Interface::Info>



lib/Function/Interface.pm
package Function::Interface;

use v5.14.0;
use warnings;

our $VERSION = "0.03";

use Carp qw(croak);
use Keyword::Simple;
use PPR;

use Function::Interface::Info;
use Function::Interface::Info::Function;
use Function::Interface::Info::Function::Param;
use Function::Interface::Info::Function::ReturnParam;

sub import {
    my $class = shift;
    my %args = @_;

    my $pkg = $args{pkg} ? $args{pkg} : caller;

    Keyword::Simple::define 'fun' => _define_interface($pkg, 'fun');
    Keyword::Simple::define 'method' => _define_interface($pkg, 'method');
}

sub unimport {
    Keyword::Simple::undefine 'fun';
    Keyword::Simple::undefine 'method';
}

sub _define_interface {
    my ($pkg, $keyword) = @_;

    return sub {
        my $ref = shift;

        my $match = _assert_valid_interface($$ref);
        my $src = _render_src($pkg, $keyword, $match);

        substr($$ref, 0, length $match->{statement}) = $src;
    }
}

sub _render_src {
    my ($pkg, $keyword, $match) = @_;

    my $src = <<"```";
BEGIN {
    Function::Interface::_register_info({
        package => '$pkg',
        keyword => '$keyword',
        subname => '$match->{subname}',
        params  => [ @{[ join ',', map {
            my $named    = $_->{named} ? 1 : 0;
            my $optional = $_->{optional} ? 1 : 0;
            qq!{ type => $_->{type}, name => '$_->{name}', named => $named, optional => $optional }!
        } @{$match->{params}} ]} ],
        return  => [ @{[ join ',', @{$match->{return}}] } ],
    });
}
```
    return $src;
}

our %metadata;
sub _register_info {
    my ($args) = @_;

    push @{$metadata{$args->{package}}} => +{
        subname => $args->{subname},
        keyword => $args->{keyword},
        params  => $args->{params},
        return  => $args->{return},
    };
}

sub info {
    my ($interface_package) = @_;
    my $info = $metadata{$interface_package} or return undef;

    Function::Interface::Info->new(
        package   => $interface_package,
        functions => [ map {
            Function::Interface::Info::Function->new(
                subname => $_->{subname},
                keyword => $_->{keyword},
                params  => [ map { _make_function_param($_) } @{$_->{params}} ],
                return  => [ map { _make_function_return_param($_) } @{$_->{return}} ],
            )
        } @{$info}],
    );
}

sub _make_function_param {
    my $param = shift;
    Function::Interface::Info::Function::Param->new(
        type     => $param->{type},
        name     => $param->{name},
        named    => $param->{named},
        optional => $param->{optional},
    )
}

sub _make_function_return_param {
    my $type = shift;
    Function::Interface::Info::Function::ReturnParam->new(
        type => $type,
    )
}

sub _assert_valid_interface {
    my $src = shift;

    $src =~ m{
        \A
        (?<statement>
            (?&PerlOWS) (?<subname>(?&PerlIdentifier))
            (?&PerlOWS) \((?<params>.*?)\)
            (?&PerlOWS) :Return\((?<return>.*?)\)
            ;
        )
        $PPR::GRAMMAR
    }sx or croak "invalid interface";

    my %match;
    $match{statement} = $+{statement};
    $match{subname} = $+{subname};
    $match{params}  = $+{params} ? _assert_valid_interface_params($+{params}) : [];
    $match{return}  = $+{return} ? _assert_valid_interface_return($+{return}) : [];

    return \%match;
}

$Function::Interface::GRAMMAR = qr{
    (?(DEFINE)
        (?<PerlType>
            (?&PerlIdentifier)
            (?: \s* \[
                \s* (?&PerlTypeParameter) \s*
                (?: , \s* (?&PerlTypeParameter) \s* )*+
            \] )?
        )

        (?<PerlTypeParameter>
            (?&PerlString)|(?&PerlVariable)|(?&PerlType)
        )
    )

    $PPR::GRAMMAR
}x;

sub _assert_valid_interface_params {
    my $src = shift;

    my @list = grep { defined } $src =~ m{
        ((?&PerlType))     \s*
        (:?) # named       \s*
        ((?&PerlVariable)) \s*
        (=?) # optional

        $Function::Interface::GRAMMAR
    }xg;

    my @params;
    while (my ($type, $named, $name, $optional) = splice @list, 0, 4) {
        push @params => {
            type     => $type,
            named    => !!$named,
            name     => $name,
            optional => !!$optional,
        }
    }

    my $regex = join '\s*,\s*', map {
        quotemeta sprintf('%s %s%s%s',
            $_->{type},
            $_->{named} ? ':' : '',
            $_->{name},
            $_->{optional} ? '=' : '',
        )
    } @params;

    croak "invalid interface params: $src"
        unless $src =~ m{ \A \s* $regex \s* \z }x;

    return \@params;
}

sub _assert_valid_interface_return {
    my $src = shift;

    my @list = grep { defined } $src =~ m{
        ((?&PerlType))
        $Function::Interface::GRAMMAR
    }xg;

    croak "invalid interface return: $src. It should be TYPELIST."
        unless $src =~ m{
            \A \s* @{[join '\s*,\s*', map { quotemeta $_ } @list]} \s* \z
        }x;

    return \@list;
}

1;
__END__

=encoding utf-8

=head1 NAME

Function::Interface - declare typed interface package

=head1 SYNOPSIS

Declare typed interface package C<IFoo>:

    package IFoo {
        use Function::Interface;
        use Types::Standard -types;

        fun hello(Str $msg) :Return(Str);

        fun add(Int $a, Int $b) :Return(Int);
    }

Implements the interface package C<IFoo>:

    package Foo {
        use Function::Interface::Impl qw(IFoo);
        use Types::Standard -types;

        fun hello(Str $msg) :Return(Str) {
            return "HELLO $msg";
        }

        fun add(Int $a, Int $b) :Return(Int) {
            return $a + $b;
        }
    }

Use the type C<ImplOf>:

    package FooService {
        use Function::Interface::Types qw(ImplOf);
        use Function::Parameters;
        use Function::Return;
        use Mouse;

        use aliased 'IFoo';

        fun greet(ImplOf[IFoo] $foo) :Return() {
            print $foo->hello;
            return;
        }
    }

    my $foo_service = FooService->new;
    my $foo = Foo->new; # implements of IFoo

    $foo_service->greet($foo);

=head1 DESCRIPTION

This module provides a typed interface.
C<Function::Interface> declares abstract functions without implementation and defines an interface package.
C<Function::Interface::Impl> checks if the abstract functions are implemented at B<compile time>.

=head2 SUPPORT

This module supports all perl versions starting from v5.14.

=head2 Declare function

C<Function::Interface> provides two new keywords, C<fun> and C<method>, for declaring abstract functions and methods with types:

    fun hello(Str $msg) :Return(Str);

    method new(Num :$x, Num :$y) :Return(Point);

The method of declaring abstract functions is the same as L<Function::Parameters> and L<Function::Return>.

=head3 declare parameters

Function arguments must always specify a variable name and type constraint, and named arguments and optional arguments can optionally be specified:

    # positional parameters
    # e.g. called `foo(1,2,3)`
    fun foo1(Int $a, Int $b, Int $c) :Return();

    # named parameters
    # e.g. called `bar(x => 123, y => 456)`
    fun foo2(Num :$x, Num :$y) :Return();

    # optional
    # e.g. called `baz()` or `baz('some')`
    fun foo3(Str $msg=) :Return();

=head3 declare return types

Specify zero or more type constraints for the function's return value:

    # zero(empty)
    fun bar1() :Return();

    # one
    fun bar2() :Return(Str);

    # two
    fun bar3() :Return(Str, Num);

=head2 requirements of type constraint

The requirements of type constraint of C<Function::Interface> is the same as for L<Function::Parameters> and L<Function::Return>.

=head1 METHODS

=head2 Function::Interface::info($interface_package)

The function C<Function::Interface::info> lets you introspect interface functions:

    # declare interface package
    package IBar {
        use Function::Interface;
        fun hello() :Return();
        fun world() :Return();
    }

    # introspect
    my $info = Function::Interface::info 'IBar';
    $info->package; # => IBar
    $info->functions; # => list of Function::Interface::Info::Function

It returns either C<undef> if it knows nothing about the interface or an object of L<Function::Interface::Info>.

=head1 SEE ALSO

L<Function::Interface::Impl>

=head1 LICENSE

Copyright (C) kfly8.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

kfly8 E<lt>kfly@cpan.orgE<gt>

=cut

